From c63c6eba5e12a9a59c9b1a5f3b2d8a1d6894f6f5 Mon Sep 17 00:00:00 2001
From: Haihao Xiang <haihao.xiang@intel.com>
Date: Fri, 2 Jul 2021 15:37:19 +0800
Subject: [PATCH 1/4] glimagesink: fix Y210 in memory:DMABuf

---
 gst-libs/gst/gl/egl/gsteglimage.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/gst-libs/gst/gl/egl/gsteglimage.c b/gst-libs/gst/gl/egl/gsteglimage.c
index 30bab22d4..d8a21be1c 100644
--- a/gst-libs/gst/gl/egl/gsteglimage.c
+++ b/gst-libs/gst/gl/egl/gsteglimage.c
@@ -91,6 +91,10 @@
 #define DRM_FORMAT_RG1616       fourcc_code('R', 'G', '3', '2')
 #endif
 
+#ifndef DRM_FORMAT_Y210
+#define DRM_FORMAT_Y210         fourcc_code('Y', '2', '1', '0')
+#endif
+
 #ifndef EGL_LINUX_DMA_BUF_EXT
 #define EGL_LINUX_DMA_BUF_EXT 0x3270
 #endif
@@ -468,10 +472,12 @@ _drm_rgba_fourcc_from_info (const GstVideoInfo * info, int plane,
   const gint rgba_fourcc = DRM_FORMAT_ABGR8888;
   const gint rgb_fourcc = DRM_FORMAT_BGR888;
   const gint rg_fourcc = DRM_FORMAT_GR88;
+  const gint rg16_fourcc = DRM_FORMAT_GR1616;
 #else
   const gint rgba_fourcc = DRM_FORMAT_RGBA8888;
   const gint rgb_fourcc = DRM_FORMAT_RGB888;
   const gint rg_fourcc = DRM_FORMAT_RG88;
+  const gint rg16_fourcc = DRM_FORMAT_RG1616;
 #endif
 
   GST_DEBUG ("Getting DRM fourcc for %s plane %i",
@@ -549,6 +555,10 @@ _drm_rgba_fourcc_from_info (const GstVideoInfo * info, int plane,
       *out_format = plane == 1 ? GST_GL_RED : GST_GL_RG;
       return plane == 1 ? rg_fourcc : DRM_FORMAT_R8;
 
+    case GST_VIDEO_FORMAT_Y210:
+      *out_format = GST_GL_RG16;
+      return rg16_fourcc;
+
     default:
       GST_ERROR ("Unsupported format for DMABuf.");
       return -1;
@@ -716,6 +726,9 @@ _drm_direct_fourcc_from_info (const GstVideoInfo * info)
     case GST_VIDEO_FORMAT_xBGR:
       return DRM_FORMAT_RGBX8888;
 
+    case GST_VIDEO_FORMAT_Y210:
+      return DRM_FORMAT_Y210;
+
     default:
       GST_INFO ("Unsupported format for direct DMABuf.");
       return -1;
-- 
2.24.3

