From 25bd9afde4fd9ec1b41d35f50b84e44b46050991 Mon Sep 17 00:00:00 2001
From: sagar patil <sagar.patil@intel.com>
Date: Sun, 12 Jul 2020 23:37:06 +0530
Subject: [PATCH] return_aad_len_from_afalg

Signed-off-by: sagar patil <sagar.patil@intel.com>
---
 engines/e_afalg.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/engines/e_afalg.c b/engines/e_afalg.c
index 8e84870..8f2d4c8 100644
--- a/engines/e_afalg.c
+++ b/engines/e_afalg.c
@@ -1746,6 +1746,7 @@ static int afalg_cipher_ctrl(EVP_CIPHER_CTX *ctx, int type, int arg, void *ptr)
      case EVP_CTRL_AEAD_TLS1_AAD:
         ectx->aad = ptr;
         ectx->aad_len = arg;
+        ret = arg;                //return aad_len for checks
         break;
       default:
             ret = -1;
-- 
2.17.1

