From c64d12adbbb06f10555abb23b4dd82d7371dadd8 Mon Sep 17 00:00:00 2001
From: Ong Boon Leong <boon.leong.ong@intel.com>
Date: Wed, 31 Jul 2019 10:28:58 +0800
Subject: [PATCH 48/52] REVERTME: add dwxpcs mocked ACPI-ID for EHL simics

Signed-off-by: Ong Boon Leong <boon.leong.ong@intel.com>
---
 drivers/net/phy/dwxpcs.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/net/phy/dwxpcs.c b/drivers/net/phy/dwxpcs.c
index ebbefb7..efbdde8 100644
--- a/drivers/net/phy/dwxpcs.c
+++ b/drivers/net/phy/dwxpcs.c
@@ -314,6 +314,7 @@ static irqreturn_t dwxpcs_interrupt(int irq, void *dev_id)
 static const struct acpi_device_id dwxpcs_acpi_match[] = {
 	{ "INTC1033" }, /* EHL Ethernet PCS */
 	{ "INTC1034" }, /* TGL Ethernet PCS */
+	{ "INTC1058" }, /* TGL Ethernet PCS */
 	{ },
 };
 
-- 
2.7.4

