From ab7b20573d4862877c9389707182a6485fcca7b5 Mon Sep 17 00:00:00 2001
From: "Yong, Jonathan" <jonathan.yong@intel.com>
Date: Mon, 30 Mar 2020 08:27:33 +0000
Subject: [PATCH 1/2] shim: fix -Werror=pointer-sign errors

upstream-status: inappropriate, interactions with new gnu-efi
Signed-off-by: Yong, Jonathan <jonathan.yong@intel.com>
---
 netboot.c | 6 +++---
 tpm.c     | 2 +-
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/netboot.c b/netboot.c
index 58babfb..ec1548e 100644
--- a/netboot.c
+++ b/netboot.c
@@ -267,7 +267,7 @@ static EFI_STATUS parseDhcp4()
 			pkt_v4 = &pxe->Mode->ProxyOffer.Dhcpv4;
 	}
 
-	INTN dir_len = strnlena(pkt_v4->BootpBootFile, 127);
+	INTN dir_len = strnlena((CHAR8 *)pkt_v4->BootpBootFile, 127);
 	INTN i;
 	UINT8 *dir = pkt_v4->BootpBootFile;
 
@@ -283,7 +283,7 @@ static EFI_STATUS parseDhcp4()
 		return EFI_OUT_OF_RESOURCES;
 
 	if (dir_len > 0) {
-		strncpya(full_path, dir, dir_len);
+		strncpya(full_path, (CHAR8 *)dir, dir_len);
 		if (full_path[dir_len-1] == '/' && template[0] == '/')
 			full_path[dir_len-1] = '\0';
 	}
@@ -334,7 +334,7 @@ EFI_STATUS FetchNetbootimage(EFI_HANDLE image_handle, VOID **buffer, UINT64 *buf
 
 try_again:
 	efi_status = pxe->Mtftp(pxe, read, *buffer, overwrite, bufsiz, &blksz,
-			      &tftp_addr, full_path, NULL, nobuffer);
+			      &tftp_addr, (UINT8 *)full_path, NULL, nobuffer);
 	if (efi_status == EFI_BUFFER_TOO_SMALL) {
 		/* try again, doubling buf size */
 		*bufsiz *= 2;
diff --git a/tpm.c b/tpm.c
index 674e69b..e4fbc09 100644
--- a/tpm.c
+++ b/tpm.c
@@ -145,7 +145,7 @@ static EFI_STATUS tpm_locate_protocol(efi_tpm_protocol_t **tpm,
 
 static EFI_STATUS tpm_log_event_raw(EFI_PHYSICAL_ADDRESS buf, UINTN size,
 				    UINT8 pcr, const CHAR8 *log, UINTN logsize,
-				    UINT32 type, CHAR8 *hash)
+				    UINT32 type, UINT8 *hash)
 {
 	EFI_STATUS efi_status;
 	efi_tpm_protocol_t *tpm;
-- 
2.20.1

