From 89ba6f1f49ad8f5875bd40fb41d27ab29840133d Mon Sep 17 00:00:00 2001
From: "Yong, Jonathan" <jonathan.yong@intel.com>
Date: Mon, 30 Mar 2020 08:30:52 +0000
Subject: [PATCH 2/2] lib/console.c: fix typo

upstream-status: inappropriate, interaction with newer gnu-efi
Signed-off-by: Yong, Jonathan <jonathan.yong@intel.com>
---
 lib/console.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/console.c b/lib/console.c
index 3aee41c..792aa58 100644
--- a/lib/console.c
+++ b/lib/console.c
@@ -445,7 +445,7 @@ static struct {
 	{  EFI_SECURITY_VIOLATION,     L"Security Violation"},
 
 	// warnings
-	{  EFI_WARN_UNKOWN_GLYPH,      L"Warning Unknown Glyph"},
+	{  EFI_WARN_UNKNOWN_GLYPH,      L"Warning Unknown Glyph"},
 	{  EFI_WARN_DELETE_FAILURE,    L"Warning Delete Failure"},
 	{  EFI_WARN_WRITE_FAILURE,     L"Warning Write Failure"},
 	{  EFI_WARN_BUFFER_TOO_SMALL,  L"Warning Buffer Too Small"},
-- 
2.20.1

