From f103800773b9cd09834ee1d3c4c8c52a37f5a494 Mon Sep 17 00:00:00 2001
From: Naina Dhiman <nainax.dhiman@intel.com>
Date: Fri, 21 Sep 2018 14:37:58 +0800
Subject: [PATCH] Makefile fix

Signed-off-by: Naina Dhiman <nainax.dhiman@intel.com>
---
 efi/Makefile | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/efi/Makefile b/efi/Makefile
index 3d7aed9..e1e37a2 100644
--- a/efi/Makefile
+++ b/efi/Makefile
@@ -23,7 +23,7 @@ CCLDFLAGS	?= -nostdlib -Wl,--warn-common \
 	-Wl,--build-id=sha1 -Wl,--hash-style=sysv \
 	$(GNUEFIDIR)/crt0-efi-$(ARCH).o
 
-define objcopy_version =
+define objcopy_version
   $(OBJCOPY) --version |
     sed -e '/^GNU objcopy/! d;
 	    : loop1; s/([^)(]*)//g;   t loop1;
@@ -130,6 +130,6 @@ endef
 install : | check_efidir_error
 install : $(FWUP).efi | $(FWUP).efi.debug $(FWUP).efi.build-id
 	$(INSTALL) -d -m 755 $(DESTDIR)$(TARGETDIR)/fw/
-	$(foreach prog,$^,$(call inst,$(prog)) ; )
+	$(foreach prog,$^,$(call inst,$(prog)))
 	$(INSTALL) -d -m 755 $(DESTDIR)$(DEBUGSOURCE)/fwupdate-$(VERSION)$(DASHRELEASE)$(DOTARCH)/efi/
 	$(foreach src,$(wildcard *.[chS]),$(INSTALL) -m 644 $(src) $(DESTDIR)$(DEBUGSOURCE)/fwupdate-$(VERSION)$(DASHRELEASE)$(DOTARCH)/efi/$(src) ;)
-- 
1.9.1

