From b7efb3934f39d39cc817c691bb2ee51cbdf6922c Mon Sep 17 00:00:00 2001
From: Vivek Kasireddy <vivek.kasireddy@intel.com>
Date: Tue, 10 Dec 2019 18:21:21 -0800
Subject: [PATCH 46/46] build: Ensure the HDR patches build with autotools as
 well

Signed-off-by: Vivek Kasireddy <vivek.kasireddy@intel.com>
---
 Makefile.am  | 113 +++++++++++++++++++++++++++++++++++----------------
 configure.ac |  27 +++++++++++-
 2 files changed, 103 insertions(+), 37 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index 5407b593..b9f0ec0c 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -39,6 +39,7 @@ AM_CPPFLAGS = 					\
 	-I$(top_builddir)/tests			\
 	-I$(top_srcdir)/shared			\
 	-I$(top_builddir)/protocol		\
+	-I$(top_srcdir)/include		\
 	-DLIBWESTON_MODULEDIR='"$(libweston_moduledir)"' \
 	-DLIBEXECDIR='"$(libexecdir)"'		\
 	-DBINDIR='"$(bindir)"'
@@ -69,13 +70,13 @@ libweston_@LIBWESTON_MAJOR@_la_SOURCES =			\
 	libweston/git-version.h				\
 	libweston/log.c					\
 	libweston/compositor.c				\
-	libweston/compositor.h				\
-	libweston/compositor-drm.h			\
-	libweston/compositor-fbdev.h			\
-	libweston/compositor-headless.h			\
-	libweston/compositor-rdp.h			\
-	libweston/compositor-wayland.h			\
-	libweston/compositor-x11.h			\
+	libweston/libweston.h				\
+	libweston/backend-drm.h			\
+	libweston/backend-fbdev.h			\
+	libweston/backend-headless.h			\
+	libweston/backend-rdp.h			\
+	libweston/backend-wayland.h			\
+	libweston/backend-x11.h			\
 	libweston/input.c				\
 	libweston/data-device.c				\
 	libweston/screenshooter.c			\
@@ -102,6 +103,8 @@ libweston_@LIBWESTON_MAJOR@_la_SOURCES =			\
 	libweston/pixel-formats.h			\
 	libweston/weston-debug.c			\
 	libweston/weston-debug.h			\
+	libweston/hdr_metadata.c			\
+	libweston/colorspace.c				\
 	shared/fd-util.h				\
 	shared/helpers.h				\
 	shared/matrix.c					\
@@ -109,6 +112,8 @@ libweston_@LIBWESTON_MAJOR@_la_SOURCES =			\
 	shared/timespec-util.h				\
 	shared/zalloc.h					\
 	shared/platform.h				\
+	shared/colorspace.c				\
+	shared/csc.c					\
 	shared/weston-egl-ext.h
 
 libweston_@LIBWESTON_MAJOR@_datadir = $(datadir)/weston/protocols
@@ -159,7 +164,7 @@ systemd_notify_la_SOURCES =			\
 	compositor/systemd-notify.c		\
 	shared/helpers.h			\
 	shared/zalloc.h				\
-	libweston/compositor.h
+	libweston/libweston.h
 endif
 
 nodist_libweston_@LIBWESTON_MAJOR@_la_SOURCES =				\
@@ -185,6 +190,10 @@ nodist_libweston_@LIBWESTON_MAJOR@_la_SOURCES =				\
 	protocol/pointer-constraints-unstable-v1-server-protocol.h      \
 	protocol/input-timestamps-unstable-v1-protocol.c		\
 	protocol/input-timestamps-unstable-v1-server-protocol.h		\
+	protocol/hdr-metadata-unstable-v1-protocol.c			\
+	protocol/hdr-metadata-unstable-v1-server-protocol.h		\
+	protocol/colorspace-unstable-v1-protocol.c			\
+	protocol/colorspace-unstable-v1-server-protocol.h		\
 	protocol/weston-touch-calibration-protocol.c			\
 	protocol/weston-touch-calibration-server-protocol.h		\
 	protocol/linux-explicit-synchronization-unstable-v1-protocol.c	\
@@ -307,23 +316,23 @@ dist_wayland_session_DATA = compositor/weston.desktop
 
 libwestonincludedir = $(includedir)/libweston-${LIBWESTON_MAJOR}
 libwestoninclude_HEADERS =			\
-	libweston/version.h			\
-	libweston/compositor.h			\
-	libweston/compositor-drm.h		\
-	libweston/compositor-fbdev.h		\
-	libweston/compositor-headless.h		\
-	libweston/compositor-rdp.h		\
-	libweston/compositor-wayland.h		\
-	libweston/compositor-x11.h		\
-	libweston/windowed-output-api.h		\
-	libweston/plugin-registry.h		\
-	libweston/timeline-object.h		\
-	shared/matrix.h				\
-	shared/config-parser.h			\
-	shared/zalloc.h
+	include/libweston/version.h			\
+	include/libweston/libweston.h			\
+	include/libweston/backend-drm.h		\
+	include/libweston/backend-fbdev.h		\
+	include/libweston/backend-headless.h		\
+	include/libweston/backend-rdp.h		\
+	include/libweston/backend-wayland.h		\
+	include/libweston/backend-x11.h		\
+	include/libweston/windowed-output-api.h		\
+	include/libweston/plugin-registry.h		\
+	include/libweston/timeline-object.h		\
+	include/libweston/matrix.h				\
+	include/libweston/config-parser.h			\
+	include/libweston/zalloc.h
 
 libwestoninclude_HEADERS +=			\
-	libweston-desktop/libweston-desktop.h
+	include/libweston-desktop/libweston-desktop.h
 
 westonincludedir = $(includedir)/weston
 westoninclude_HEADERS = compositor/weston.h
@@ -346,13 +355,17 @@ gl_renderer_la_CFLAGS =				\
 	$(LIBDRM_CFLAGS)			\
 	$(AM_CFLAGS)
 gl_renderer_la_SOURCES =			\
-	libweston/gl-renderer.h			\
-	libweston/gl-renderer.c			\
+	libweston/renderer-gl/gl-renderer.h	\
+	libweston/renderer-gl/gl-renderer-private.h	\
+	libweston/renderer-gl/gl-renderer.c	\
+	libweston/renderer-gl/gl-shaders.c	\
 	libweston/vertex-clipping.c		\
 	libweston/vertex-clipping.h		\
 	libweston/linux-sync-file.c		\
 	libweston/linux-sync-file.h		\
 	libweston/linux-sync-file-uapi.h	\
+	shared/colorspace.c			\
+	shared/csc.c				\
 	shared/helpers.h
 endif
 
@@ -373,7 +386,7 @@ x11_backend_la_CFLAGS =				\
 	$(X11_COMPOSITOR_CFLAGS)
 x11_backend_la_SOURCES = 			\
 	libweston/compositor-x11.c		\
-	libweston/compositor-x11.h		\
+	libweston/backend-x11.h		\
 	shared/helpers.h
 endif
 
@@ -405,10 +418,11 @@ drm_backend_la_CFLAGS =				\
 	$(AM_CFLAGS)
 drm_backend_la_SOURCES =			\
 	libweston/compositor-drm.c		\
-	libweston/compositor-drm.h		\
+	libweston/backend-drm.h		\
 	$(INPUT_BACKEND_SOURCES)		\
 	shared/helpers.h			\
 	shared/timespec-util.h			\
+	libweston/drm-hdr-metadata.c		\
 	libweston/libbacklight.c		\
 	libweston/libbacklight.h
 
@@ -455,7 +469,7 @@ wayland_backend_la_CFLAGS =			\
 	$(AM_CFLAGS)
 wayland_backend_la_SOURCES = 					\
 	libweston/compositor-wayland.c				\
-	libweston/compositor-wayland.h				\
+	libweston/backend-wayland.h				\
 	shared/helpers.h
 nodist_wayland_backend_la_SOURCES =				\
 	protocol/fullscreen-shell-unstable-v1-protocol.c	\
@@ -474,7 +488,7 @@ headless_backend_la_LIBADD =			\
 headless_backend_la_CFLAGS = $(COMPOSITOR_CFLAGS) $(AM_CFLAGS)
 headless_backend_la_SOURCES = 			\
 	libweston/compositor-headless.c		\
-	libweston/compositor-headless.h		\
+	libweston/backend-headless.h		\
 	shared/helpers.h
 endif
 
@@ -497,7 +511,7 @@ fbdev_backend_la_CFLAGS =			\
 	$(AM_CFLAGS)
 fbdev_backend_la_SOURCES =			\
 	libweston/compositor-fbdev.c		\
-	libweston/compositor-fbdev.h		\
+	libweston/backend-fbdev.h		\
 	shared/helpers.h			\
 	$(INPUT_BACKEND_SOURCES)
 endif
@@ -516,7 +530,7 @@ rdp_backend_la_CFLAGS =				\
 	$(AM_CFLAGS)
 rdp_backend_la_SOURCES = 			\
 	libweston/compositor-rdp.c		\
-	libweston/compositor-rdp.h		\
+	libweston/backend-rdp.h		\
 	shared/helpers.h
 endif
 
@@ -560,7 +574,7 @@ spring_tool_SOURCES =				\
 	libweston/animation.c			\
 	shared/matrix.c				\
 	shared/matrix.h				\
-	libweston/compositor.h
+	libweston/libweston.h
 
 if BUILD_CLIENTS
 
@@ -678,6 +692,24 @@ weston_simple_dmabuf_drm_LDADD = $(SIMPLE_DMABUF_DRM_CLIENT_LIBS) \
 	libshared.la
 endif
 
+demo_clients += weston-simple-hdr-video-gbm
+weston_simple_hdr_video_gbm_SOURCES =			\
+	clients/simple-hdr-video-gbm.c
+nodist_weston_simple_hdr_video_gbm_SOURCES =		\
+	protocol/xdg-shell-protocol.c			\
+	protocol/xdg-shell-client-protocol.h		\
+	protocol/linux-dmabuf-unstable-v1-protocol.c \
+	protocol/linux-dmabuf-unstable-v1-client-protocol.h \
+	protocol/hdr-metadata-unstable-v1-protocol.c \
+	protocol/hdr-metadata-unstable-v1-client-protocol.h \
+	protocol/colorspace-unstable-v1-protocol.c \
+	protocol/colorspace-unstable-v1-client-protocol.h
+weston_simple_hdr_video_gbm_CFLAGS = $(AM_CFLAGS) $(SIMPLE_HDR_VIDEO_CLIENT_CFLAGS) $(CLIENT_CFLAGS) $(PANGO_CFLAGS)
+weston_simple_hdr_video_gbm_LDADD = $(SIMPLE_HDR_VIDEO_CLIENT_LIBS) \
+	$(LIBDRM_PLATFORM_INTEL_LIBS)     \
+	libtoytoolkit.la	\
+	libshared.la
+
 if BUILD_SIMPLE_DMABUF_V4L_CLIENT
 demo_clients += weston-simple-dmabuf-v4l
 weston_simple_dmabuf_v4l_SOURCES = clients/simple-dmabuf-v4l.c
@@ -904,7 +936,7 @@ nodist_weston_info_SOURCES =					\
 	protocol/xdg-output-unstable-v1-protocol.c		\
 	protocol/xdg-output-unstable-v1-client-protocol.h
 weston_info_LDADD = $(WESTON_INFO_LIBS) libshared.la
-weston_info_CFLAGS = $(AM_CFLAGS) $(CLIENT_CFLAGS)
+weston_info_CFLAGS = $(AM_CFLAGS) $(CLIENT_CFLAGS) -I$(top_srcdir)/include
 
 weston_debug_SOURCES = 					\
 	clients/weston-debug.c				\
@@ -983,6 +1015,10 @@ BUILT_SOURCES +=					\
 	protocol/input-timestamps-unstable-v1-client-protocol.h		\
 	protocol/xdg-output-unstable-v1-protocol.c			\
 	protocol/xdg-output-unstable-v1-client-protocol.h		\
+	protocol/hdr-metadata-unstable-v1-protocol.c			\
+	protocol/hdr-metadata-unstable-v1-client-protocol.h		\
+	protocol/colorspace-unstable-v1-protocol.c			\
+	protocol/colorspace-unstable-v1-client-protocol.h		\
 	protocol/linux-explicit-synchronization-unstable-v1-protocol.c	\
 	protocol/linux-explicit-synchronization-unstable-v1-client-protocol.h
 
@@ -1039,6 +1075,8 @@ desktop_shell_la_CPPFLAGS =			\
 	-I$(top_builddir)/protocol		\
 	-I$(top_srcdir)/shared			\
 	-I$(top_builddir)/libweston		\
+	-I$(top_srcdir)/include		\
+	-I$(top_srcdir)/include/libweston		\
 	-I$(top_srcdir)/libweston		\
 	-I$(top_builddir)/desktop-shell		\
 	-DMODULEDIR='"$(moduledir)"'		\
@@ -1068,6 +1106,7 @@ fullscreen_shell_la_CPPFLAGS =			\
 	-I$(top_builddir)/protocol		\
 	-I$(top_srcdir)/shared			\
 	-I$(top_builddir)/libweston		\
+	-I$(top_srcdir)/include		\
 	-I$(top_srcdir)/libweston
 
 fullscreen_shell_la_LDFLAGS = -module -avoid-version
@@ -1167,6 +1206,8 @@ xwayland_la_CPPFLAGS =				\
 	-I$(top_srcdir)/shared			\
 	-I$(top_builddir)/libweston		\
 	-I$(top_srcdir)/libweston		\
+	-I$(top_srcdir)/include/libweston		\
+	-I$(top_srcdir)/include		\
 	-I$(top_builddir)/xwayland		\
 	-DMODULEDIR='"$(moduledir)"'		\
 	-DLIBEXECDIR='"$(libexecdir)"'
@@ -1192,7 +1233,7 @@ xwayland_la_SOURCES =				\
 	xwayland/hash.h				\
 	shared/helpers.h
 
-libwestoninclude_HEADERS += xwayland/xwayland-api.h
+libwestoninclude_HEADERS += include/libweston/xwayland-api.h
 
 endif
 
@@ -1208,7 +1249,7 @@ libshared_la_CPPFLAGS = \
 	-DDATADIR='"$(datadir)"' \
 	$(AM_CPPFLAGS)
 
-libshared_la_CFLAGS = $(AM_CFLAGS) $(COMPOSITOR_CFLAGS)
+libshared_la_CFLAGS = $(AM_CFLAGS) $(COMPOSITOR_CFLAGS) -I$(top_srcdir)/include
 
 libshared_la_SOURCES =				\
 	shared/config-parser.c			\
@@ -1540,7 +1581,7 @@ matrix_test_SOURCES =				\
 	tests/matrix-test.c			\
 	shared/matrix.c				\
 	shared/matrix.h
-matrix_test_CPPFLAGS = -DUNIT_TEST
+matrix_test_CPPFLAGS = -DUNIT_TEST -I$(top_srcdir)/include
 matrix_test_LDADD = -lm $(CLOCK_GETTIME_LIBS)
 
 if ENABLE_IVI_SHELL
diff --git a/configure.ac b/configure.ac
index c05ad011..2e0787c7 100644
--- a/configure.ac
+++ b/configure.ac
@@ -455,6 +455,31 @@ if ! test "x$enable_simple_dmabuf_drm_client" = "xno"; then
 fi
 AM_CONDITIONAL(BUILD_SIMPLE_DMABUF_DRM_CLIENT, test "x$enable_simple_dmabuf_drm_client" = "xyes")
 
+AC_ARG_ENABLE(simple-hdr-video-client,
+              AS_HELP_STRING([--disable-simple-hdr-video-client],
+                             [do not build the simple dmabuf drm client]),,
+              enable_simple_hdr_video_client="auto")
+if ! test "x$enable_simple_hdr_video_client" = "xno"; then
+  PKG_CHECK_MODULES(SIMPLE_HDR_VIDEO_CLIENT, [wayland-client libdrm libavutil libavcodec libavformat gbm egl glesv2], [have_simple_hdr_video_libs=yes],
+		    [have_simple_hdr_video_libs=no])
+
+  PKG_CHECK_MODULES(LIBDRM_PLATFORM_INTEL, [libdrm_intel],
+      AC_DEFINE([HAVE_LIBDRM_INTEL], [1], [Build intel dmabuf client]) have_simple_hdr_video_client=yes,
+      [true])
+
+  if test "x$have_simple_hdr_video_client" != "xyes" -o \
+	  "x$have_simple_hdr_video_libs" = "xno" && \
+     test "x$enable_simple_hdr_video_client" = "xyes"; then
+    AC_MSG_ERROR([DRM dmabuf client explicitly enabled, but none of libdrm_{intel,freedreno,etnaviv} found])
+  fi
+
+  if test "x$have_simple_hdr_video_client" = "xyes" -a "x$have_simple_hdr_video_libs" = "xyes"; then
+    enable_simple_hdr_video_client="yes"
+  fi
+fi
+AM_CONDITIONAL(BUILD_SIMPLE_HDR_VIDEO_CLIENT, test "x$enable_simple_hdr_video_client" = "xyes")
+
+
 AC_ARG_ENABLE(simple-dmabuf-v4l-client,
               AS_HELP_STRING([--disable-simple-dmabuf-v4l-client],
                              [do not build the simple dmabuf v4l client]),,
@@ -734,7 +759,7 @@ if test "x$enable_systemd_notify" = "xyes"; then
   PKG_CHECK_MODULES(SYSTEMD_DAEMON, [libsystemd])
 fi
 
-AC_CONFIG_FILES([Makefile libweston/version.h compositor/weston.pc])
+AC_CONFIG_FILES([Makefile include/libweston/version.h compositor/weston.pc])
 
 # AC_CONFIG_FILES needs the full name when running autoconf, so we need to use
 # libweston_abi_version here, and outside [] because of m4 quoting rules
-- 
2.21.0

