From 295a00a59277fd30d78b916c0e49629d31eaeccb Mon Sep 17 00:00:00 2001
From: "Soon, Thean Siew" <thean.siew.soon@intel.com>
Date: Fri, 4 Aug 2023 22:10:04 +0000
Subject: [PATCH] va: Add scaling and composition pipeline flags

Signed-off-by: Soon, Thean Siew <thean.siew.soon@intel.com>
---
 sys/va/gstvafilter.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/sys/va/gstvafilter.c b/sys/va/gstvafilter.c
index 98e433887c..5465cb360b 100644
--- a/sys/va/gstvafilter.c
+++ b/sys/va/gstvafilter.c
@@ -61,6 +61,7 @@ struct _GstVaFilter
   GstVideoOrientationMethod orientation;
 
   guint32 scale_method;
+  guint32 pipeline_flags;
 
   gboolean crop_enabled;
 
@@ -919,7 +920,12 @@ gst_va_filter_set_scale_method (GstVaFilter * self, guint32 method)
   g_return_val_if_fail (GST_IS_VA_FILTER (self), FALSE);
 
   GST_OBJECT_LOCK (self);
+  guint32 pipeline_flags = 0;
+  if (method == VA_FILTER_SCALING_FAST) {
+    pipeline_flags = VA_PROC_PIPELINE_FAST;
+  }
   self->scale_method = method;
+  self->pipeline_flags = pipeline_flags;
   GST_OBJECT_UNLOCK (self);
 
   return TRUE;
@@ -1608,6 +1614,7 @@ _create_pipeline_buffer (GstVaFilter * self, GstVaSample * src,
     .input_color_properties = self->input_color_properties,
     .output_color_properties = self->output_color_properties,
     .filter_flags = self->scale_method,
+    .pipeline_flags = self->pipeline_flags,
     /* output to SDR */
     .output_hdr_metadata = NULL,
   };
@@ -1791,8 +1798,10 @@ gst_va_filter_compose (GstVaFilter * self, GstVaComposeTransaction * tx)
       .output_region = &sample->output_region,
       .output_background_color = 0xff000000,
       .filter_flags = self->scale_method,
+      .pipeline_flags = self->pipeline_flags,
     };
     /* *INDENT-ON* */
+
     GST_OBJECT_UNLOCK (self);
 
     /* only send blend state when sample is not fully opaque */
-- 
2.34.1

