From 86af8ec8a0a3a6927ac99c49777a7f6a57f9292f Mon Sep 17 00:00:00 2001
From: Lim Siew Hoon <siew.hoon.lim@intel.com>
Date: Mon, 14 Aug 2023 13:24:10 +0800
Subject: [PATCH 1/2] msdk: Use local cflags for configuration

Signed-off-by: Lim Siew Hoon <siew.hoon.lim@intel.com>
---
 sys/msdk/meson.build | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/sys/msdk/meson.build b/sys/msdk/meson.build
index b73370cda..85f04899d 100644
--- a/sys/msdk/meson.build
+++ b/sys/msdk/meson.build
@@ -28,6 +28,7 @@ msdk_dep = []
 use_msdk = false
 use_onevpl = false
 onevpl_extra_args = []
+msdk_extra_args = []
 
 extra_args = [
   '-DGST_USE_UNSTABLE_API',
@@ -121,7 +122,7 @@ endif
 
 if use_onevpl or cxx.has_header('mfxvp9.h', args: '-I' + mfx_incdir)
   msdk_sources += [ 'gstmsdkvp9dec.c' ]
-  cdata.set('USE_MSDK_VP9_DEC', 1)
+  msdk_extra_args += ['-DUSE_MSDK_VP9_DEC=1']
 endif
 
 # Usually MFX_VERSION 1026+ is required to support raw VP9 stream, however Fedora 30 has MFX_VERSION==1026
@@ -141,7 +142,7 @@ have_mfx_ver126 = cc.compiles(mfx_ver126_check_code,
 
 if have_mfx_ver126
   msdk_sources += [ 'gstmsdkvp9enc.c' ]
-  cdata.set('USE_MSDK_VP9_ENC', 1)
+  msdk_extra_args += ['-DUSE_MSDK_VP9_ENC=1']
 endif
 
 mfx_ver134_check_code = '''
@@ -156,7 +157,7 @@ have_mfx_ver134 = cc.compiles(mfx_ver134_check_code,
 
 if have_mfx_ver134
   msdk_sources += [ 'gstmsdkav1dec.c' ]
-  cdata.set('USE_MSDK_AV1_DEC', 1)
+  msdk_extra_args += ['-DUSE_MSDK_AV1_DEC=1']
 endif
 
 mfx_ver205_check_code = '''
@@ -171,7 +172,7 @@ have_mfx_ver205 = cc.compiles(mfx_ver205_check_code,
 
 if use_onevpl and have_mfx_ver205
   msdk_sources += [ 'gstmsdkav1enc.c' ]
-  cdata.set10('USE_MSDK_AV1_ENC', 1)
+  msdk_extra_args += ['-DUSE_MSDK_AV1_ENC=1']
 endif
 
 if host_machine.system() == 'windows'
@@ -199,7 +200,7 @@ if msdk_deps_found
 
   gstmsdktag = library('gstmsdk',
     msdk_sources,
-    c_args : gst_plugins_bad_args + extra_args + onevpl_extra_args,
+    c_args : gst_plugins_bad_args + extra_args +  msdk_extra_args + onevpl_extra_args,
     include_directories : [configinc, mfx_inc],
     dependencies : [gstbase_dep, gstvideo_dep, gstpbutils_dep, gstallocators_dep, gstcodecparsers_dep, mfx_dep, msdk_deps],
     install : true,
-- 
2.43.2

