From 26adcc4f8ee5bcc7701450d107a3d66067888180 Mon Sep 17 00:00:00 2001
From: "Soon, Thean Siew" <thean.siew.soon@intel.com>
Date: Fri, 4 Aug 2023 22:10:04 +0000
Subject: [PATCH 94/98] va: Add scaling and composition pipeline flags

Signed-off-by: Soon, Thean Siew <thean.siew.soon@intel.com>
---
 sys/va/gstvafilter.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/sys/va/gstvafilter.c b/sys/va/gstvafilter.c
index 0eb1d7659115..48d477793c83 100644
--- a/sys/va/gstvafilter.c
+++ b/sys/va/gstvafilter.c
@@ -61,6 +61,7 @@ struct _GstVaFilter
   GstVideoOrientationMethod orientation;
 
   guint32 scale_method;
+  guint32 pipeline_flags;
 
   gboolean crop_enabled;
 
@@ -924,7 +925,13 @@ gst_va_filter_set_scale_method (GstVaFilter * self, guint32 method)
   g_return_val_if_fail (GST_IS_VA_FILTER (self), FALSE);
 
   GST_OBJECT_LOCK (self);
+  guint32 pipeline_flags = 0;
+  if (method == VA_FILTER_SCALING_FAST) {
+    pipeline_flags = VA_PROC_PIPELINE_FAST;
+  }
+
   self->scale_method = method;
+  self->pipeline_flags = pipeline_flags;
   GST_OBJECT_UNLOCK (self);
 
   return TRUE;
@@ -1613,6 +1620,7 @@ _create_pipeline_buffer (GstVaFilter * self, GstVaSample * src,
     .input_color_properties = self->input_color_properties,
     .output_color_properties = self->output_color_properties,
     .filter_flags = self->scale_method,
+    .pipeline_flags = self->pipeline_flags,
     /* output to SDR */
     .output_hdr_metadata = NULL,
   };
@@ -1789,6 +1797,7 @@ gst_va_filter_compose (GstVaFilter * self, GstVaComposeTransaction * tx)
       .surface_region = &sample->input_region,
       .output_region = &sample->output_region,
       .filter_flags = self->scale_method,
+      .pipeline_flags = self->pipeline_flags,
     };
     /* *INDENT-ON* */
     GST_OBJECT_UNLOCK (self);
-- 
2.43.0

