From 8f4d0787d33760bd340c0fb19978eda0c76823c0 Mon Sep 17 00:00:00 2001
From: "Cheah, Vincent Beng Keat" <vincent.beng.keat.cheah@intel.com>
Date: Thu, 17 Aug 2023 19:37:22 +0800
Subject: [PATCH 95/98] va: compositor: Disable tiling for i965 for DMA RGB
 formats

---
 sys/va/gstvacompositor.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/sys/va/gstvacompositor.c b/sys/va/gstvacompositor.c
index ccbc4763f5ba..ad4779400483 100644
--- a/sys/va/gstvacompositor.c
+++ b/sys/va/gstvacompositor.c
@@ -731,7 +731,8 @@ gst_va_compositor_decide_allocation (GstAggregator * agg, GstQuery * query)
       goto bail;
     }
 
-    if (gst_caps_is_dmabuf (caps) && GST_VIDEO_INFO_IS_RGB (&info)) {
+    if (GST_VA_DISPLAY_IS_IMPLEMENTATION (self->display, INTEL_I965) &&
+        gst_caps_is_dmabuf (caps) && GST_VIDEO_INFO_IS_RGB (&info)) {
       usage_hint = VA_SURFACE_ATTRIB_USAGE_HINT_GENERIC;
     } else {
       usage_hint = va_get_surface_usage_hint (self->display,
-- 
2.43.0

