From 0190a2a737fd75d42e7d67b68a1d758811e6f003 Mon Sep 17 00:00:00 2001
From: "Cheam, Xin Jie" <xin.jie.cheam@intel.com>
Date: Thu, 11 Jul 2024 18:09:41 +0800
Subject: [PATCH 97/98] msdkdec : WA set allocation cap's width,height without
 aligned value

---
 sys/msdk/gstmsdkdec.c | 8 ++------
 1 file changed, 2 insertions(+), 6 deletions(-)

diff --git a/sys/msdk/gstmsdkdec.c b/sys/msdk/gstmsdkdec.c
index 5e64321deed2..605d18d52727 100644
--- a/sys/msdk/gstmsdkdec.c
+++ b/sys/msdk/gstmsdkdec.c
@@ -844,12 +844,8 @@ gst_msdkdec_set_src_caps (GstMsdkDec * thiz, gboolean need_allocation)
 
   if (need_allocation) {
     /* Find allocation width and height */
-    width =
-        GST_ROUND_UP_16 (thiz->param.mfx.FrameInfo.Width ? thiz->param.mfx.
-        FrameInfo.Width : GST_VIDEO_INFO_WIDTH (&thiz->output_state->info));
-    height =
-        GST_ROUND_UP_32 (thiz->param.mfx.FrameInfo.Height ? thiz->param.mfx.
-        FrameInfo.Height : GST_VIDEO_INFO_HEIGHT (&thiz->output_state->info));
+    width = GST_VIDEO_INFO_WIDTH (&thiz->output_state->info);
+    height = GST_VIDEO_INFO_HEIGHT (&thiz->output_state->info);
 
     /* set allocation width and height in allocation_caps,
      * which may or may not be similar to the output_state caps */
-- 
2.43.0

