From 84654e8587362c53a31e756f264ab76d4296fbe7 Mon Sep 17 00:00:00 2001
From: "Cheah, Vincent Beng Keat" <vincent.beng.keat.cheah@intel.com>
Date: Thu, 17 Aug 2023 19:37:22 +0800
Subject: [PATCH 230/232] va: compositor: Disable tiling for i965 for DMA RGB
 formats

---
 sys/va/gstvacompositor.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/sys/va/gstvacompositor.c b/sys/va/gstvacompositor.c
index aae85a0ac..053c5c615 100644
--- a/sys/va/gstvacompositor.c
+++ b/sys/va/gstvacompositor.c
@@ -732,7 +732,8 @@ gst_va_compositor_decide_allocation (GstAggregator * agg, GstQuery * query)
   }
 
   if (!allocator) {
-    if (gst_caps_is_dmabuf (caps) && GST_VIDEO_INFO_IS_RGB (&info))
+    if (GST_VA_DISPLAY_IS_IMPLEMENTATION (self->display, INTEL_I965) &&
+        gst_caps_is_dmabuf (caps) && GST_VIDEO_INFO_IS_RGB (&info))
       usage_hint = VA_SURFACE_ATTRIB_USAGE_HINT_GENERIC;
     if (!(allocator = gst_va_compositor_allocator_from_caps (self, caps)))
       return FALSE;
-- 
2.40.1

