From 7f3cbb48600b05b934285c90f3329b991c08052d Mon Sep 17 00:00:00 2001
From: Dongwon Kim <dongwon.kim@intel.com>
Date: Wed, 29 Jun 2022 12:04:06 -0700
Subject: [PATCH] virtio-gpu: not deleting resources when resetting GPU

Signed-off-by: Dongwon Kim <dongwon.kim@intel.com>
---
 hw/display/virtio-gpu.c | 7 -------
 1 file changed, 7 deletions(-)

diff --git a/hw/display/virtio-gpu.c b/hw/display/virtio-gpu.c
index 397ae3ef8e..7b22c20868 100644
--- a/hw/display/virtio-gpu.c
+++ b/hw/display/virtio-gpu.c
@@ -1347,13 +1347,8 @@ void virtio_gpu_device_realize(DeviceState *qdev, Error **errp)
 void virtio_gpu_reset(VirtIODevice *vdev)
 {
     VirtIOGPU *g = VIRTIO_GPU(vdev);
-    struct virtio_gpu_simple_resource *res, *tmp;
     struct virtio_gpu_ctrl_command *cmd;
 
-    QTAILQ_FOREACH_SAFE(res, &g->reslist, next, tmp) {
-        virtio_gpu_resource_destroy(g, res);
-    }
-
     while (!QTAILQ_EMPTY(&g->cmdq)) {
         cmd = QTAILQ_FIRST(&g->cmdq);
         QTAILQ_REMOVE(&g->cmdq, cmd, next);
@@ -1366,8 +1361,6 @@ void virtio_gpu_reset(VirtIODevice *vdev)
         g->inflight--;
         g_free(cmd);
     }
-
-    virtio_gpu_base_reset(VIRTIO_GPU_BASE(vdev));
 }
 
 static void
-- 
2.20.1

