From 6dab60cd3f49d4e37a90acbb98690426c42cfab0 Mon Sep 17 00:00:00 2001
From: Otavio Salvador <otavio@ossystems.com.br>
Date: Tue, 1 Jul 2014 17:37:31 -0300
Subject: [PATCH] build: Allow CC and prefix to be overriden

Upstream-Status: Pending

Signed-off-by: Otavio Salvador <otavio@ossystems.com.br>
---
 makefile | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/makefile b/makefile
index d09a4a9..c88caf5 100644
--- a/makefile
+++ b/makefile
@@ -18,7 +18,7 @@
 KBUILD_OUTPUT =
 
 DEBUG	=
-CC	= $(CROSS_COMPILE)gcc
+CC	?= $(CROSS_COMPILE)gcc
 VER     = -DVER=$(version)
 CFLAGS	= -Wall $(VER) $(incdefs) $(DEBUG) $(EXTRA_CFLAGS)
 LDLIBS	= -lm -lrt $(EXTRA_LDFLAGS)
@@ -46,7 +46,7 @@ OBJECTS	+= snmp4lptp.o
 snmplib	:= $(shell net-snmp-config --netsnmp-agent-libs)
 endif
 
-prefix	= /usr/local
+prefix	?= /usr/local
 sbindir	= $(prefix)/sbin
 mandir	= $(prefix)/man
 man8dir	= $(mandir)/man8
-- 
2.17.0

