From 281b20b7cd2b35289f653718e5118625687dc61a Mon Sep 17 00:00:00 2001
From: "Soon, Thean Siew" <thean.siew.soon@intel.com>
Date: Wed, 17 Jan 2018 16:35:21 +0800
Subject: [PATCH] Temporary Workaround to remove X11 dependency

---
 CMakeLists.txt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index c554e5c..37411cb 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -30,7 +30,7 @@ option (MFX_SINK "Build MSDK sink plugin." ON)
 CMAKE_DEPENDENT_OPTION(WITH_WAYLAND "Enable Wayland support"
     ON "MFX_SINK" OFF)
 CMAKE_DEPENDENT_OPTION(WITH_X11  "Enable X11 support"
-    ON "MFX_SINK" OFF)
+    OFF "MFX_SINK" OFF)

 CMAKE_DEPENDENT_OPTION(USE_WAYLAND_RENDERER "Build sink plugin with Wayland backend"
     ON "WITH_WAYLAND" OFF)
--
1.9.1

