From 84b70fa3ad1673e644b7af5847f6c56eb11c1437 Mon Sep 17 00:00:00 2001
From: johnmach <john.machado@intel.com>
Date: Tue, 19 Apr 2022 17:42:13 +0530
Subject: [PATCH 2/2] Add RPL-P DIDs (#98)

---
 Source/inc/common/igfxfmid.h | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/Source/inc/common/igfxfmid.h b/Source/inc/common/igfxfmid.h
index bd275032cdae..d45bddcf0ce6 100644
--- a/Source/inc/common/igfxfmid.h
+++ b/Source/inc/common/igfxfmid.h
@@ -1804,6 +1804,14 @@ typedef enum __NATIVEGTTYPE
 #define DEV_ID_56B0                             0x56B0
 #define DEV_ID_56B1                             0x56B1
 
+// RPL-P
+#define DEV_ID_A7A0                             0xA7A0
+#define DEV_ID_A7A1                             0xA7A1
+#define DEV_ID_A7A8                             0xA7A8
+#define DEV_ID_A7A9                             0xA7A9
+#define DEV_ID_A720                             0xA720
+#define DEV_ID_A721                             0xA721
+
 // ADL-N
 #define DEV_ID_46D0                            0x46D0
 #define DEV_ID_46D1                            0x46D1
-- 
2.34.1

