diff --git a/driver/linux/Makefile b/driver/linux/Makefile
index 1bba7e7..fffb87b 100644
--- a/driver/linux/Makefile
+++ b/driver/linux/Makefile
@@ -72,64 +72,24 @@ else
 endif
 
 else
-KDIR := /lib/modules/$(shell uname -r)/build
+KDIR := $(KERNEL_SRC)
 PWD  := $(shell pwd)
 
-ifneq ($(SIG_FILE),)
-FULL_SIG_FILE = $(shell realpath $(SIG_FILE))
-export FULL_SIG_FILE
-endif
-
-SGX_LE_SIGNING_KEY_PATH := sgx_signing_key.pem
-SGX_LE_PUBLIC_KEY_PATH := sgx_public_key.pem
-SGX_LE_SIGNING_MATERIAL := signing_material.bin
-
-SIGNING_KEY_PATH := $(shell realpath $(SGX_LE_SIGNING_KEY_PATH))
-PUBLIC_KEY_PATH := $(shell realpath $(SGX_LE_PUBLIC_KEY_PATH))
-SIGNING_MATERIAL := $(shell realpath $(SGX_LE_SIGNING_MATERIAL))
-
-export SIGNING_KEY_PATH
-export PUBLIC_KEY_PATH
-export SIGNING_MATERIAL
-
-sign:
-	$(MAKE) -C $(KDIR) SUBDIRS=$(PWD) CFLAGS_MODULE="-I$(PWD) -I$(PWD)/include" modules LE_ACTION=SIGN
-
-gendata:
-	$(MAKE) -C $(KDIR) SUBDIRS=$(PWD) CFLAGS_MODULE="-I$(PWD) -I$(PWD)/include" modules LE_ACTION=GENDATA
-
-usesig:
-	$(MAKE) -C $(KDIR) SUBDIRS=$(PWD) CFLAGS_MODULE="-I$(PWD) -I$(PWD)/include" modules LE_ACTION=USESIG
-
-endif
-
-ifeq ($(LE_ACTION)$(FULL_SIG_FILE),USESIG)
- $(error On usesig build - SIG_FILE must point to the signature input file)
 endif
 
 KBUILD_AFLAGS += -I$(PWD)/le 
 
-$(eval $(call config_filename,INTEL_SGX_SIGNING_KEY))
-
-
-$(SIGNING_KEY_PATH):
-	$(Q)openssl genrsa -3 -out $(SIGNING_KEY_PATH) 3072
+all:
+	$(MAKE) -C $(KDIR) M=$(PWD) CFLAGS_MODULE="-I$(PWD) -I$(PWD)/include" modules
 
-$(PUBLIC_KEY_PATH): $(SIGNING_KEY_PATH)
-	$(Q)openssl rsa -in $(SIGNING_KEY_PATH) -outform PEM -pubout -out $(PUBLIC_KEY_PATH)
-
-
-ifeq ($(LE_ACTION),SIGN)
-SGX_LE_KEY_DEP := $(SIGNING_KEY_PATH)
-else ifeq ($(LE_ACTION),USESIG)
-SGX_LE_KEY_DEP := $(PUBLIC_KEY_PATH)
-endif
+modules_install:
+	$(MAKE) -C $(KDIR) M=$(PWD) CFLAGS_MODULE="-I$(PWD) -I$(PWD)/include" modules_install
 
 $(obj)/dummy.o: $(SIGNING_MATERIAL)
 $(obj)/sgx_le_proxy_piggy.o: $(SGX_LE_KEY_DEP) $(obj)/le/sgx_le_proxy
 $(obj)/le/sgx_le_proxy $(SIGNING_MATERIAL): FORCE
 	$(Q)$(MAKE) $(build)=$(obj)/le $@
- 
+
 clean:
 	rm -vrf *.o *.ko *.order *.symvers *.mod.c .tmp_versions .*.cmd $(SIGNING_MATERIAL)
 	$(Q)$(MAKE) -C le clean
