diff --git a/driver/linux/sgx_page_cache.c b/driver/linux/sgx_page_cache.c
index 3a75dc9..a256c47 100644
--- a/driver/linux/sgx_page_cache.c
+++ b/driver/linux/sgx_page_cache.c
@@ -94,8 +94,13 @@ static atomic_t sgx_nr_free_pages = ATOMIC_INIT(0);
 static struct task_struct *ksgxswapd_tsk;
 static DECLARE_WAIT_QUEUE_HEAD(ksgxswapd_waitq);
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 3, 0))
+static int sgx_test_and_clear_young_cb(pte_t *ptep,
+				       unsigned long addr, void *data)
+#else
 static int sgx_test_and_clear_young_cb(pte_t *ptep, pgtable_t token,
 				       unsigned long addr, void *data)
+#endif
 {
 	pte_t pte;
 	int ret;
