From 7cd6a6b57019e2fd9b003e2cca8a292f249e5d75 Mon Sep 17 00:00:00 2001
From: "Yew, Chang Ching" <chang.ching.yew@intel.com>
Date: Thu, 27 Jan 2022 18:20:55 +0800
Subject: [PATCH] Adding RPL-S dev id

Signed-off-by: Chew, Tong Liang <tong.liang.chew@intel.com>
---
 dispatcher/linux/device_ids.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/dispatcher/linux/device_ids.h b/dispatcher/linux/device_ids.h
index 0f008497..615c3b76 100644
--- a/dispatcher/linux/device_ids.h
+++ b/dispatcher/linux/device_ids.h
@@ -44,6 +44,9 @@ enum eMFXHWType {
 
     MFX_HW_JSL = 0x1500001,
     MFX_HW_EHL = 0x1500002,
+
+    MFX_HW_TGL_LP = 0x1600000,
+    MFX_HW_RPL_S  = MFX_HW_TGL_LP + 6,
 };
 
 enum eMFXGTConfig { MFX_GT_UNKNOWN = 0, MFX_GT1 = 1, MFX_GT2 = 2, MFX_GT3 = 3, MFX_GT4 = 4 };
-- 
2.34.1

