From a9f4dfa580fa2fd0e4e6113378ad01788b0a970c Mon Sep 17 00:00:00 2001
From: Dongwon Kim <dongwon.kim@intel.com>
Date: Mon, 22 Aug 2022 11:46:17 -0700
Subject: [PATCH 6/8] ui/gtk: hide the memu bar when full-screening

Signed-off-by: Dongwon Kim <dongwon.kim@intel.com>
---
 ui/gtk.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/ui/gtk.c b/ui/gtk.c
index 849c7375cd..67e0d0ded5 100644
--- a/ui/gtk.c
+++ b/ui/gtk.c
@@ -2408,6 +2408,9 @@ static void gtk_display_init(DisplayState *ds, DisplayOptions *opts)
                 for (; i < s->nb_vcs; i++) {
                     win = s->vc[i].window ? s->vc[i].window : s->window;
                     if (opts->has_full_screen && opts->full_screen) {
+                        if (gtk_widget_is_visible(s->menu_bar)) {
+                            gtk_widget_hide(s->menu_bar);
+                        }
                         gtk_window_fullscreen_on_monitor(
                             GTK_WINDOW(win),
                             gdk_display_get_default_screen(window_display),
-- 
2.20.1

