From 0b5e06c8c76060100463067bc81c4d4c34c142f2 Mon Sep 17 00:00:00 2001
From: Dongwon Kim <dongwon.kim@intel.com>
Date: Fri, 21 Jun 2024 06:45:01 +0000
Subject: [PATCH 37/40] gtk: untabifying the primary window as well

Signed-off-by: Dongwon Kim <dongwon.kim@intel.com>
---
 ui/gtk.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ui/gtk.c b/ui/gtk.c
index c0d9b5755e75..307ec8e3324f 100644
--- a/ui/gtk.c
+++ b/ui/gtk.c
@@ -1730,7 +1730,7 @@ static void gd_menu_untabify(GtkMenuItem *item, void *opaque)
         gtk_check_menu_item_set_active(GTK_CHECK_MENU_ITEM(s->grab_item),
                                        FALSE);
     }
-    if (!vc->window) {
+    if (!vc->window || (vc->window == s->window)) {
         gd_tab_window_create(vc);
     }
 
-- 
2.34.1

